### PR TITLE
A new `broadcastspend` command

### LINKED ISSUE #58: daemon: a new `broadcast` command

### LINKED ISSUE #66: The `test_update_spend` test may pass an invalid PSBT to `updatespend`
I could reproduce this by stress-testing the functional test suite (running each test a dozen time, and running 20 of them in parallel). Here is the trace:
```
minisafed = <test_framework.minisafed.Minisafed object at 0x7f016ab015b0>, bitcoind = <test_framework.bitcoind.Bitcoind object at 0x7f016a235190>                                                                                                                                                                                                                          
                                                                                                                                                                                                                                                                                                                                                                           
    def test_update_spend(minisafed, bitcoind):                                                                                                                                                                                                                                                                                                                            
        # Start by creating a Spend PSBT                                                                                                                                                                                                                                                                                                                                   
        addr = minisafed.rpc.getnewaddress()["address"]                                                                                                                                                                                                                                                                                                                    
        bitcoind.rpc.sendtoaddress(addr, 0.2567)                                                                                                                                                                                                                                                                                                                           
        wait_for(lambda: len(minisafed.rpc.listcoins()["coins"]) > 0)                                                                                                                                                                                                                                                                                                      
        outpoints = [c["outpoint"] for c in minisafed.rpc.listcoins()["coins"]]                                                                                                                                                                                                                                                                                            
        destinations = {                                                                                                                                                                                                                                                                                                                                                   
            bitcoind.rpc.getnewaddress(): 200_000,                                                                                                                                                                                                                                                                                                                         
        }                                                                                                                                                                                                                                                                                                                                                                  
        res = minisafed.rpc.createspend(outpoints, destinations, 6)                                                                                                                                                                                                                                                                                                        
        assert "psbt" in res                                                                                                                                                                                                                                                                                                                                               
                                                                                                                                                                                                                                                                                                                                                                           
        # Now update it                                                                                                                                                                                                                                                                                                                                                    
        assert len(minisafed.rpc.listspendtxs()["spend_txs"]) == 0                                                                                                                                                                                                                                                                                                         
        minisafed.rpc.updatespend(res["psbt"])                                                                                                                                                                                                                                                                                                                             
        list_res = minisafed.rpc.listspendtxs()["spend_txs"]                                                                                                                                                                                                                                                                                                               
        assert len(list_res) == 1                                                                                                                                                                                                                                                                                                                                          
        assert list_res[0]["psbt"] == res["psbt"]                                                                                                                                                                                                                                                                                                                          
                                                                                                                                                                                                                                                                                                                                                                           
        # Keep a copy for later.                                                                                                                                                                                                                                                                                                                                           
        psbt_no_sig = PSBT()                                                                                                                                                                                                                                                                                                                                               
        psbt_no_sig.deserialize(res["psbt"])                                                                                                                                                                                                                                                                                                                               
                                                                                                                                                                                                                                                                                                                                                                           
        # We can add a signature and update it                                                                                                                                                                                                                                                                                                                             
        psbt_sig_a = PSBT()                                                                                                                                                                                                                                                                                                                                                
        psbt_sig_a.deserialize(res["psbt"])                                                                                                                                                                                                                                                                                                                                
        dummy_pk_a = bytes.fromhex(                                                                                                                                                                                                                                                                                                                                        
            "0375e00eb72e29da82b89367947f29ef34afb75e8654f6ea368e0acdfd92976b7c"                                                                                                                                                                                                                                                                                           
        )                                                                                                                                                                                                                                                                                                                                                                  
        dummy_sig_a = bytes.fromhex(                                                                                                                                                                                                                                                                                                                                       
            "304402202b925395cfeaa0171a7a92982bb4891acc4a312cbe7691d8375d36796d5b570a0220378a8ab42832848e15d1aedded5fb360fedbdd6c39226144e527f0f1e19d5398"                                                                                                                                                                                                                 
        )                                                                                                                                                                                                                                                                                                                                                                  
        psbt_sig_a.inputs[0].partial_sigs[dummy_pk_a] = dummy_sig_a                                                                                                                                                                                                                                                                                                        
        psbt_sig_a_ser = psbt_sig_a.serialize()                                                                                                                                                                                                                                                                                                                            
>       minisafed.rpc.updatespend(psbt_sig_a_ser)                                                                                                                                                                                                                                                                                                                          
                                                                                                                                                                                                                                                                                                                                                                           
tests/test_rpc.py:204:                                                                                                                                                                                                                                                                                                                                                     
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _                                                                                                                                                                                        
tests/test_framework/utils.py:192: in wrapper                                                                                                                                                                                                                                                                                                                              
    return self.call(name, params=args or kwargs)                                                                                                                                                                                                                                                                                                                          
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                                                                                                           
self = <test_framework.utils.UnixDomainSocketRpc object at 0x7f016ab01970>, method = 'updatespend'                                                                                                                                                                                                                                                                         
params = ('cHNidP8BAH0CAAAAAdZ3tAhuLjlmIiMbOglxoMs9U7RhO44UVGT5YIwCJfkSAAAAAAD/////AkANAwAAAAAAFgAU9lUPboqXX6UK33mwhwwCN4RT+u40...AbP80ILAnT1JcB4/AUQhA7dGO9gcJ01qHlitjRlyEFmo6j1WefLxWHfB/4BDHen1rHNkdqkU/g/83EjSWTTSvK0AMjPiYu8Sx5KIrQPo/QCyaAAAAA==',)                                                                                                                  
                                                                                                                                                                                                                                                                                                                                                                           
    def call(self, method, params={}):                                                                                                                                                                                                                                                                                                                                     
        self.logger.debug(f"Calling {method} with params {params}")                                                                                                                                                                                                                                                                                                        
                                                                                                                                                                                    
        # FIXME: we open a new socket for every readobj call...                                                                                                                     
        sock = UnixSocket(self.socket_path)                                                                                                                                         
        msg = json.dumps(                                                                                                                                                           
            {                                                                                                                                                                       
                "jsonrpc": "2.0",                                                                                                                                                   
                "id": 0,                                                                                                                                                            
                "method": method,                                                                                                                                                   
                "params": params,                                                                                                                                                   
            }                                                                                                                                                                       
        )                                                                                                                                                                           
        sock.sendall(msg.encode() + b"\n")                                                                                                                                          
        this_id = self.next_id                                                                                                                                                      
        resp = self._readobj(sock)                                                                                                                                                  
                                                                                                                                                                                    
        self.logger.debug(f"Received response for {method} call: {resp}")                                                                                                           
        if "id" in resp and resp["id"] != this_id:                                                                                                                                  
            raise ValueError(                                                                                                                                                       
                "Malformed response, id is not {}: {}.".format(this_id, resp)                                                                                                       
            )                                                                                                                                                                       
        sock.close()                                                                                                                                                                
                                                                                                                                                                                    
        if not isinstance(resp, dict):                                                                                                                                              
            raise ValueError(                                                                                                                                                       
                f"Malformed response, response is not a dictionary: {resp}"                                                                                                         
            )                                                                                                                                                                       
        elif "error" in resp:                                                                                                                                                       
>           raise RpcError(method, params, resp["error"])                                                                                                                           
E           test_framework.utils.RpcError: RPC call failed: method: updatespend, params: ('cHNidP8BAH0CAAAAAdZ3tAhuLjlmIiMbOglxoMs9U7RhO44UVGT5YIwCJfkSAAAAAAD/////AkANAwAAAAAAFgAU9lUPboqXX6UK33mwhwwCN4RT+u40oIQBAAAAACIAIDXvbJZFritqW7xET1twUHgAP6xkIiN0F2WJYqH2KBRlAAAAAAABAStwsYcBAAAAACIAIDXvbJZFritqW7xET1twUHgAP6xkIiN0F2WJYqH2KBRlAQgCSDBFAiEAuZYvKjORaXD2S7xBZqI4
xbNnrcUA3iQQLta9GPM1VC0CIFhSaycEgPUypZctLC9VdncdjhyWAbP80ILAnT1JcB4/AUQhA7dGO9gcJ01qHlitjRlyEFmo6j1WefLxWHfB/4BDHen1rHNkdqkU/g/83EjSWTTSvK0AMjPiYu8Sx5KIrQPo/QCyaAAAAA==',), error: {'code': -32602, 'message': "Invalid params: Invalid 'feerate' parameter."}
```

(I fixed the "invalid 'feerate' parameter" typo in #15)

### PR DESCRIPTION
**Pull request by darosior**
This implements a new command to broadcast an existing Spend transaction.

Fixes #58.
Fixes #66.

### CODE REVIEW

File: src/commands/mod.rs [hunk 1]
```diff
@@ -436,6 +445,30 @@ impl DaemonControl {
         let mut db_conn = self.db.connection();
         db_conn.delete_spend(txid);
     }
+
+    /// Finalize and broadcast this stored Spend transaction.
+    pub fn broadcast_spend(&self, txid: &bitcoin::Txid) -> Result<(), CommandError> {
+        let mut db_conn = self.db.connection();
+
+        log::debug!("A");
```
Comment (edouardparis) [inline src/commands/mod.rs:453]: forgot to delete it

File: src/commands/mod.rs [hunk 2]
```diff
@@ -436,6 +445,30 @@ impl DaemonControl {
...
+        log::debug!("B");
+        miniscript::psbt::finalize(&mut spend_psbt, &self.secp)
+            .map_err(|e| CommandError::SpendFinalization(e.to_string()))?;
+
+        // Then, broadcast it (or try to, we never know if we are not going to hit an
+        // error at broadcast time).
+        let final_tx = spend_psbt.extract_tx();
+        match self.bitcoin.broadcast_tx(&final_tx) {
+            Ok(()) => Ok(()),
+            Err(BitcoinError::Broadcast(e)) => Err(CommandError::TxBroadcast(e)),
```
Comment (edouardparis) [inline src/commands/mod.rs:469]: you can use map_err or do a custom `From`  for CommandError

**Review by edouardparis — APPROVED** [review]

File: src/commands/mod.rs (hunk 1 above)
Comment (darosior) [reply src/commands/mod.rs:453]: Thanks

File: src/commands/mod.rs (hunk 2 above)
Comment (darosior) [reply src/commands/mod.rs:469]: That's intentional. The advantage of using `match` here is that when/if we introduce a new `BitcoinError` variant in the future that should not be reachable from there, we can treat is as such.

**Comment by darosior** [comment]
Rebased after the merge of #15

**Comment by darosior** [comment]
Arf `test_broadcast_spend` is flaky. I'll investigate tomorrow.

```
minisafed = <test_framework.minisafed.Minisafed object at 0x7fb2f4351190>, bitcoind = <test_framework.bitcoind.Bitcoind object at 0x7fb2f54746a0>                                                                                                                                                                                                                          
                                                                                                                                                                                                                                                                                                                                                                           
    def test_broadcast_spend(minisafed, bitcoind):                                                                                                                                                                                                                                                                                                                         
        # Create a new coin and a spending tx for it.                                                                                                                                                                                                                                                                                                                      
        addr = minisafed.rpc.getnewaddress()["address"]                                                                                                                                                                                                                                                                                                                    
        bitcoind.rpc.sendtoaddress(addr, 0.2567)                                                                                                                                                                                                                                                                                                                           
        wait_for(lambda: len(minisafed.rpc.listcoins()["coins"]) > 0)                                                                                                                                                                                                                                                                                                      
        outpoints = [c["outpoint"] for c in minisafed.rpc.listcoins()["coins"]]                                                                                                                                                                                                                                                                                            
        destinations = {                                                                                                                                                                                                                                                                                                                                                   
            bitcoind.rpc.getnewaddress(): 200_000,                                                                                                                                                                                                                                                                                                                         
        }                                                                                                                                                                                                                                                                                                                                                                  
        res = minisafed.rpc.createspend(outpoints, destinations, 6)                                                                                                                                                                                                                                                                                                        
        psbt = PSBT()                                                                                                                                                                                                                                                                                                                                                      
        psbt.deserialize(res["psbt"])                                                                                                                                                                                                                                                                                                                                      
        txid = psbt.tx.txid().hex()                                                                                                                                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                                                                                                           
        # We can't broadcast an unknown Spend                                                                                                                                                                                                                                                                                                                              
        with pytest.raises(RpcError, match="Unknown spend transaction.*"):                                                                                                                                                                                                                                                                                                 
            minisafed.rpc.broadcastspend(txid)                                                                                                                                                                                                                                                                                                                             
        minisafed.rpc.updatespend(res["psbt"])                                                                                                                                                                                                                                                                                                                             
                                                                                                                                                                                                                                                                                                                                                                           
        # We can't broadcast an unsigned transaction                                                                                                                                                                                                                                                                                                                       
        with pytest.raises(RpcError, match="Failed to finalize the spend transaction.*"):                                                                                                                                                                                                                                                                                  
            minisafed.rpc.broadcastspend(txid)                                                                                                                                                                                                                                                                                                                             
        signed_psbt = minisafed.sign_psbt(psbt)                                                                                                                                                                                                                                                                                                                            
>       minisafed.rpc.updatespend(signed_psbt.serialize())

tests/test_rpc.py:261:                                                                    
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  
tests/test_framework/utils.py:195: in wrapper                                             
    return self.call(name, params=args or kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  

self = <test_framework.utils.UnixDomainSocketRpc object at 0x7fb2f4351b20>, method = 'updatespend'                                                                                   
params = ('cHNidP8BAH0CAAAAAWT9Q1gB0ePaAgVBlX6zOsZLnOeapqx/xngDQzJqyVg6AAAAAAD/////AkANAwAAAAAAFgAU20c6GELCAS0osr7v7Kcr8R4SNt00...KLEKmzYFwVSthRfEg4ABRCEDay6oCeh2eyEHG3z0xblwhdj9iZvtxhIXAqt2x1FqzbWsc2R2qRT+D/zcSNJZNNK8rQAyM+Ji7xLHkoitA+j9ALJoAAAA',)

    def call(self, method, params={}):                                                    
        self.logger.debug(f"Calling {method} with params {params}")
                                                                                          
        # FIXME: we open a new socket for every readobj call...
        sock = UnixSocket(self.socket_path)                                               
        msg = json.dumps(                                                                 
            {                                                                             
                "jsonrpc": "2.0",                                                         
                "id": 0,                                                                  
                "method": method,                                                         
                "params": params,                                                         
            }                                                                             
        )                                                                                 
        sock.sendall(msg.encode() + b"\n")                                                
        this_id = self.next_id                                                            
        resp = self._readobj(sock)                                                        
                                                                                          
        self.logger.debug(f"Received response for {method} call: {resp}")
        if "id" in resp and resp["id"] != this_id:
            raise ValueError(                                                             
                "Malformed response, id is not {}: {}.".format(this_id, resp)
            )                                                                             
        sock.close()                                                                      
                                                                                          
        if not isinstance(resp, dict):                                                    
            raise ValueError(                                                             
                f"Malformed response, response is not a dictionary: {resp}"
            )                                                                             
        elif "error" in resp:                                                             
>           raise RpcError(method, params, resp["error"])
E           test_framework.utils.RpcError: RPC call failed: method: updatespend, params: ('cHNidP8BAH0CAAAAAWT9Q1gB0ePaAgVBlX6zOsZLnOeapqx/xngDQzJqyVg6AAAAAAD/////AkANAwAAAAAAFgAU20c6GELCAS0osr7v7Kcr8R4SNt00oIQBAAAAACIAII12Cl5HKhYPUEfIrJW7aEZWwf/8fvM3KO0PK00EeJ87AAAAAAABAStwsYcBAAAAACIAII12Cl5HKhYPUEfIrJW7aEZWwf/8fvM3KO0PK00EeJ87AQgCRzBEAiBwjkMGJr256sJ/lS5HwpXFf8cGTQ98ZFAAs+jYw/1M7wIgYa9XIWWBYisI1Y8Ejo7dMa2EKLEKmzYFwVSthRfEg4ABRCEDay6oCeh2eyEHG3z0xblwhdj9iZvtxhIXAqt2x1FqzbWsc2R2qRT+D/zcSNJZNNK8rQAyM+Ji7xLHkoitA+j9ALJoAAAA',), error: {'code': -32602, 'message': "Invalid params: Invalid 'psbt' parameter."}
```

**Comment by darosior** [comment]
Phew. Could finally track it down. But in order to fix it i had to rewrite our Python implementation of PSBTs.

**Comment by darosior** [comment]
ACK b89401e5835b94f7e5341028357b11db69f8e37d

**Pull request merged**